crates/re_renderer/src/renderer/depth_cloud.rs

```diff
@@ -0,0 +1,356 @@
+//! Since there's no widespread support for bindless textures, this requires one bind group and one
+//! draw call per texture.
+//! This is a pretty heavy shader though, so the overhead is minimal.
+//!
+//! The vertex shader backprojects the depth texture using the user-specified intrinsics, and then
+//! behaves pretty much exactly like our point cloud renderer (see [`point_cloud.rs`]).
+
+use smallvec::smallvec;
+use std::num::NonZeroU32;
+
+use crate::{
+    allocator::create_and_fill_uniform_buffer_batch,
+    include_file,
+    resource_managers::ResourceManagerError,
+    view_builder::ViewBuilder,
+    wgpu_resources::{
+        BindGroupDesc, BindGroupEntry, BindGroupLayoutDesc, GpuBindGroup, GpuBindGroupLayoutHandle,
+        GpuRenderPipelineHandle, PipelineLayoutDesc, RenderPipelineDesc, ShaderModuleDesc,
+        TextureDesc,
+    },
+};
+
+use super::{
+    DrawData, DrawOrder, FileResolver, FileSystem, RenderContext, Renderer, SharedRendererData,
+    WgpuResourcePools,
+};
+
+// ---
+
+mod gpu_data {
+    // - Keep in sync with mirror in depth_cloud.wgsl.
+    // - See `DepthCloud` for documentation.
+    #[repr(C, align(256))]
+    #[derive(Clone, Copy, bytemuck::Pod, bytemuck::Zeroable)]
+    pub struct DepthCloudInfoUBO {
+        pub world_from_obj: crate::wgpu_buffer_types::Mat4,
+        pub intrinsics: crate::wgpu_buffer_types::Mat3,
+        pub radius_scale: crate::wgpu_buffer_types::F32RowPadded,
+
+        pub end_padding: [crate::wgpu_buffer_types::PaddingRow; 16 - 9],
+    }
+}
+
+/// The raw data from a depth texture.
+///
+/// This is either `u16` or `f32` values; in both cases the data will be uploaded to the shader
+/// as-is.
+/// For `u16`s, this results in a `Depth16Unorm` texture, otherwise an `R32Float`.
+///
+/// The shader assumes that this is normalized, linear, non-flipped depth using the camera
+/// position as reference point (not the camera plane!).
+//
+// TODO(cmc): support more depth data types.
+// TODO(cmc): expose knobs to linearize/normalize/flip/cam-to-plane depth.
+#[derive(Debug, Clone)]
+pub enum DepthCloudDepthData {
+    U16(Vec<u16>),
+    F32(Vec<f32>),
+}
+
+impl DepthCloudDepthData {
+    fn as_bytes(&self) -> &[u8] {
+        match self {
+            DepthCloudDepthData::U16(data) => bytemuck::cast_slice(data),
+            DepthCloudDepthData::F32(data) => bytemuck::cast_slice(data),
+        }
+    }
+}
+
+pub struct DepthCloud {
+    pub world_from_obj: glam::Mat4,
+
+    /// The intrinsics of the camera used for the projection.
+    ///
+    /// Only supports pinhole cameras at the moment.
+    pub intrinsics: glam::Mat3,
+
+    /// The scale to apply to the radii of the backprojected points.
+    pub radius_scale: f32,
+
+    /// The dimensions of the depth texture in pixels.
+    pub depth_dimensions: glam::UVec2,
+
+    /// The actual data from the depth texture.
+    ///
+    /// See [`DepthCloudDepthData`] for more information.
+    pub depth_data: DepthCloudDepthData,
```

Comment:
As mentioned above, I think we should take the 2d texture handle here and have the upload be happening elsewhere